PermissionsScraper/PermissionsScraper/Triggers/DescriptionsScraper.cs

```diff
@@ -0,0 +1,191 @@
+
+                    log.LogInformation($"Exiting function DescriptionsScraper. Time: {DateTime.UtcNow}");
+
+                    return;
+                }
+                log.LogInformation($"No permissions descriptions update required. Exiting function 'DescriptionsScraper'. Time: {DateTime.UtcNow}");
+            }
+            catch (Exception ex)
+            {
+                log.LogInformation($"Exception occurred: {ex.InnerException?.Message ?? ex.Message}. Time: {DateTime.UtcNow}");
+            }
+        }
+
+        /// <summary>
+        /// Retrieves and populates permissions descriptions from a Service Principal.
+        /// </summary>
+        /// <param name="config">The application configuration settings.</param>
+        /// <param name="result">The JSON response of the permissions and their descriptions retrieved from the Service Prinicpal.</param>
+        /// <param name="version">The version of the API from which to fetch the scopes descriptions from the Service Principal.</param>
+        private static void PopulateScopesDescriptions(PermissionsAppConfig config,
+                                                       AuthenticationResult result,
+                                                       string version)
+        {
+            string webApiUrl = $"{config.ApiUrl}{version}/serviceprincipals?$filter=appId eq '{config.ServicePrincipalId}'";
+            var spJson = ProtectedApiCallHelper
+                    .CallWebApiAsync(webApiUrl, result.AccessToken)
+                    .GetAwaiter().GetResult();
+
+            if (string.IsNullOrEmpty(spJson))
+            {
+                throw new ArgumentNullException(nameof(spJson), $"The call to fetch the Service Principal returned empty data. URL: {webApiUrl} ");
+            }
+
+            var spJsonResponse = PermissionsFormatHelper.FormatServicePrincipalResponse(spJson, config);
+
+            // Retrieve the top level scope dictionary
+            var spValue = JsonConvert.DeserializeObject<JObject>(spJsonResponse).Value<JArray>(config.TopLevelDictionaryName);
+
+            if (spValue == null)
+            {
+                throw new ArgumentNullException(nameof(config.TopLevelDictionaryName), $"Attempt to retrieve the top-level dictionary returned empty data." +
+                    $"Name: {config.TopLevelDictionaryName}");
+            }
+
+            /* Fetch permissions defined in the second level dictionary(ies),
+             * e.g. appRoles, oauth2PermissionScopes --> 2nd level dictionary keys
+             */
+            foreach (var scopeName in config.ScopesNames)
+            {
+                // Retrieve all scopes descriptions for a given 2nd level dictionary retrieved from the Service Principal
+                var scopeDescriptions = spValue.First.Value<JArray>(scopeName)?.ToObject<List<Dictionary<string, object>>>();
```

Comment:
`scopeDescriptions` can be `null` after this line.

Reply:
Resolved